bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/log/EquinoxLogWriter.java

```diff
@@ -315,9 +318,10 @@ private synchronized void log(FrameworkLogEntry logEntry) {
 			} catch (Exception e2) {
 				System.err.println("An exception occurred while logging to the console:");//$NON-NLS-1$
 				e2.printStackTrace(System.err);
+			} finally {
+				// ensure that the error stream writer is closed
+				closeFile();
 			}
-		} finally {
```

Comment:
Do we ensure the writer is closed when the framework stops?
I think we need an option to preserve the old behavior also.

Reply:
> Do we ensure the writer is closed when the framework stops?

Actually, i thought the already existing `close` method would be called. But indeed it is not - it is called by the wrapping `FrameworkLog` but that service is manually registered, and it is never actually invoked.  Instead, i have now added calls to `close` to the `EquinoxLogWriter` instances  in the `stop` method of `EquinoxLogServices` - that works, but it would have been nicer if the `FrameworkLog` service had a lifecycle method for disposal (as with DS @Deactivate, but the service is manually registered).

> I think we need an option to preserve the old behavior also.

What kind of configuration mechanism would you suggest (VM Argument)? And what would be your expected default behavior (default = old behavior or default = new behavior)?

Reply:
> i have now added calls to close to the EquinoxLogWriter instances in the stop method of EquinoxLogServices - that works, but it would have been nicer if the FrameworkLog service had a lifecycle method for disposal (as with DS @deactivate, but the service is manually registered).

Thanks, I think this is fine for a solution here.  The framework cannot depend on DS for service management.

Reply:
@tjwatson thanks, i have now added a new member `PROP_CLOSE_LOG_FILE_EAGERLY` for the new property. I've added it to `EquinoxLogWriter` instead of `EquinoxConfiguration` as all the other related log file properties are also declared in there.